devito/passes/iet/misc.py

```diff
@@ -76,47 +87,75 @@ def relax_incr_dimensions(iet, **kwargs):
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+            assert i.direction is Forward
+            if i.dim.parent in proc_parents_max and i.symbolic_size == i.dim.parent.step:
+                # In hierarchical blocking (BLOCKLEVELS > 1) a parent dimension may
+                # have already been processed as an 'inner' iteration. Since in
+                # hierarchical blocking, block sizes in lower levels always perfectly
+                # divide block sizes of upper levels we can use parent's iteration
+                # maximum.
+                iter_max = proc_parents_max[i.dim.parent]
+            else:
+                # Candidate 1: symbolic_max of current iteration
+                # Most of the cases pass though this code:
+                # Candidates for iteration's maximum are:
+                # Candidate 1: symbolic_max of current iteration
+                # e.g. `i.symbolic_max = x0_blk0 + x0_blk0_size`
+                symbolic_max = i.symbolic_max
+
+                # Candidate 2: maximum of the root iteration
+                # Candidate 2a: Usualy it is the max of parent/root
+                # dimension, e.g. `x_M`
+                root_max = roots_max[i.dim.root]
+
+                # Candidate 2b:
+                # the symbolic_size of an Iteration may exceed the size of a parent's
+                # block size (e.g. after CIRE passes)
+                # e.g.
+                # `i.dim.parent.step = x0_blk1_size`
+                # `i.symbolic_size = x0_blk1_size + 4`
+                # For this case, proper margin should be allowed
+                # in order not to omit loop iterations.
+                lb_margin = i.symbolic_min - i.dim.symbolic_min
+                size_margin = i.symbolic_size - i.dim.parent.step
+                ub_margin = i.dim.parent.symbolic_max + size_margin + lb_margin
+
+                # So Candidate 2 is the maximum of the
+                # root's max (2a) and the current iteration's required max (2b)
+                # and instead of `x_M` we may need `x_M + 1` or `x_M + 2`
+
+                # So candidate 2 is
+                # e.g. `domain_max = Max(x_M + 1, x_M)``
+                domain_max = MAX(ub_margin, root_max)
+
+                # Finally the iteration's maximum is the minimum of the
+                # candidates 1 and 2
+                # e.g. `iter_max = Min(x0_blk0 + x0_blk0_size, domain_max)``
+                iter_max = MIN(symbolic_max, domain_max)
+
+            # Store the selected maximum of this iteration's dimension for
+            # possible reference in case of children iterations
+            # Usually encountered in subdims and hierarchical blocking
+            proc_parents_max[i.dim] = iter_max
+
+            mapper[i] = i._rebuild(limits=(i.symbolic_min, iter_max, i.step))
+
+    iet = Transformer(mapper, nested=True).visit(iet)
+
+    headers = [('MIN(a,b)', ('(((a) < (b)) ? (a) : (b))')),
```

Comment:
probably cleaner something along the lines of `'%s(a, b)' % MIN.name`

Reply:
ok